bob-access/src/authenticator/basic.rs

```diff
@@ -1,65 +1,132 @@
-use std::{collections::HashMap, net::IpAddr};
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{RwLock, Arc},
+    time::Duration,
+};
 
-use crate::{credentials::{Credentials, CredentialsKind}, AuthenticationType, error::Error, permissions::Permissions};
+use crate::{credentials::{DeclaredCredentials, DCredentialsResolveGuard, RequestCredentials, CredentialsKind}, AuthenticationType, error::Error, permissions::Permissions};
 
 use super::{users_storage::UsersStorage, Authenticator};
 
 use sha2::{Digest, Sha512};
 
+use tokio::net::lookup_host;
+
+#[inline]
+async fn lookup(hostname: &str) -> Option<Vec<SocketAddr>> {
+    lookup_host(hostname).await
+    .ok()
+    .map(|addr| addr.collect())
+}
+
+type NodesCredentials = HashMap<String, DCredentialsResolveGuard>;
+
 #[derive(Debug, Default, Clone)]
 pub struct Basic<Storage: UsersStorage> {
     users_storage: Storage,
-    nodes: HashMap<IpAddr, Vec<Credentials>>,
+    nodes: Arc<RwLock<NodesCredentials>>,
+    resolve_sleep_period_ms: u64,
 }
 
 impl<Storage: UsersStorage> Basic<Storage> {
-    pub fn new(users_storage: Storage) -> Self {
+    pub fn new(users_storage: Storage, resolve_sleep_period_ms: u64) -> Self {
         Self {
             users_storage,
-            nodes: HashMap::new(),
+            nodes: Arc::new(RwLock::new(HashMap::new())),
+            resolve_sleep_period_ms,
         }
     }
 
+    fn node_creds_ok(creds: &HashMap<String, DeclaredCredentials>) -> bool {
+        creds.values()
+            .all(|cred| cred.validate_internode())
+    }
+
     pub fn set_nodes_credentials(
         &mut self,
-        nodes: HashMap<IpAddr, Vec<Credentials>>,
+        mut nodes: HashMap<String, DeclaredCredentials>,
     ) -> Result<(), Error> {
-        if nodes
-            .values()
-            .all(|creds|
-                creds
-                    .iter()
-                    .all(|cred|
-                        cred.ip().is_some() &&
-                        cred.kind().map(|k| k.is_internode()) == Some(true)))
-        {
-            self.nodes = nodes;
+        if Self::node_creds_ok(&nodes) {
+            let mut nodes_creds = self.nodes.write().expect("nodes credentials lock");
+            for (nodename, cred) in nodes.drain() {
+                let guard = DCredentialsResolveGuard::new(cred.clone(), self.resolve_sleep_period_ms);
+                nodes_creds.insert(nodename, guard);
+                if cred.ip().is_empty() && cred.hostname().is_some() {
+                    self.spawn_resolver(cred);
```

Comment:
You should set resolving status to `InProgress` before spawning the resolver thread